momaland/envs/ingenious/ingenious_base.py

```diff
@@ -0,0 +1,427 @@
+        else:
+            # when color type is not 6( like 1-5), the number of combinations could be divided by NUM_TILES(120)
+            self.tiles_bag = int(NUM_TILES / len(diff_color_combinations + same_color_combinations)) * (
+                diff_color_combinations + same_color_combinations
+            )
+        # Shuffle the tiles bag
+        self.random.shuffle(self.tiles_bag)
+
+    def set_action_index(self, index):
+        """Apply the corresponding action for the given index on the board."""
+        """If selected actions is not a legal move, return False"""
+        if self.masked_action[index] == 0:
+            assert "illegal move:choose a masked action(0)"
+            return False
+        if (index not in self.first_round_pos) and self.first_round:
+            assert "illegal move: choose a action legally in first round"
+            return False
+        """Hex Coordinate: h1,h2 ;  Tile to play: card"""
+        h1, h2, card = self.action_index_map[index]
+        agent_i = self.agent_selector
+        agent = self.agents[agent_i]
+        if card >= len(self.p_tiles[agent]):
+            assert "illegal move: choosing tile out of hand(happening after ingenious)"
+            return False
+        """Extract the certain tile (color1 , color2) as (c1,c2)"""
+        c1, c2 = self.p_tiles[agent][card]
+        # Translate Hex Coordinate to Offset Coordinate(x,y)
+        x1, y1 = Hex2ArrayLocation(h1, self.board_size)
+        x2, y2 = Hex2ArrayLocation(h2, self.board_size)
+        flag = False
+        for item in self.p_tiles[agent]:
+            # print(item)
+            if (c1, c2) == item:
+                self.p_tiles[agent].remove(item)
+                flag = True
+                break
+            if (c2, c1) == item:
+                self.p_tiles[agent].remove(item)
+                flag = True
+                break
+        if not flag:
+            assert "illegal move: set the tile to the coordinate unsuccessfully"
+            return False
+        """Update the mask_action list after the action"""
+        self.legal_move.remove(index)
+        self.board_array[x1][y1] = c1
+        self.board_array[x2][y2] = c2
+        self.exclude_action(h1)
+        self.exclude_action(h2)
+        skip_flag = False
+        """Update score through checking 5 neighboring directions for h1 and h2 independently"""
+        point = 0
+        for i in range(0, 6):
+            h1n = hex_neighbor(h1, i)
+            if h1n == h2:
+                continue
+            # print(h1n)
+            while h1n in self.board_hex:
+                x, y = Hex2ArrayLocation(h1n, self.board_size)
+                # print(x,y,h1n)
+                if self.board_array[x][y] == c1:
+                    point += 1
+                else:
+                    break
+                h1n = hex_neighbor(h1n, i)
+                # print(h1n)
+        self.score[agent][c1] += point
```

Comment:
I would create a separate method that scores a single piece of a tile and re-use it to avoid code duplication. Something like this (haven't tested the method in code btw)

```
def calculate_score_for_piece(self, start_hex, other_hex, color):
    point = 0
    for i in range(0, 6):
        neighbor_hex = hex_neighbor(start_hex, i)
        if neighbor_hex == other_hex:
            continue
        while neighbor_hex in self.board_hex:
            x, y = Hex2ArrayLocation(neighbor_hex, self.board_size)
            if self.board_array[x][y] == color:
                point += 1
            else:
                break
            neighbor_hex = hex_neighbor(neighbor_hex, i)
    return point

# You can then re-use the method to score both pieces of the tile without code duplication.
self.score[agent][c1] += self.calculate_score_for_piece(h1, h2, c1)
if self.score[agent][c1] > self.limitation_score:
    skip_flag = True
    self.score[agent][c1] = 0

self.score[agent][c2] += self.calculate_score_for_piece(h2, h1, c2)
if self.score[agent][c2] > self.limitation_score:
    skip_flag = True
    self.score[agent][c2] = 0
```

Reply:
agree.